File: airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/AirbyteMessageBufferedWriter.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.internal;

import io.airbyte.protocol.models.AirbyteMessage;
import java.io.IOException;

/**
 * Interface for writing airbyte messages. Base interface that the versioned writers build upon.
 */
public interface AirbyteMessageBufferedWriter {

  void write(AirbyteMessage message) throws IOException;

  void flush() throws IOException;

  void close() throws IOException;

}

File: airbyte-container-orchestrator/src/main/kotlin/io/airbyte/container/orchestrator/worker/io/AirbyteMessageBufferedWriter.kt
/*
 * Copyright (c) 2020-2026 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.container.orchestrator.worker.io

import io.airbyte.commons.protocol.AirbyteMessageVersionedMigrator
import io.airbyte.commons.protocol.serde.AirbyteMessageSerializer
import io.airbyte.config.ConfiguredAirbyteCatalog
import io.airbyte.protocol.models.v0.AirbyteMessage
import java.io.BufferedWriter
import java.util.Optional

/**
 * Write protocol objects in a specified version.
 *
 * @param <T> type of protocol object.
</T> */
class AirbyteMessageBufferedWriter<T : Any>(
  private val writer: BufferedWriter,
  private val serializer: AirbyteMessageSerializer<T>,
  private val migrator: AirbyteMessageVersionedMigrator<T>,
  private val configuredAirbyteCatalog: Optional<ConfiguredAirbyteCatalog>,
) {
  fun write(message: AirbyteMessage) {
    val downgradedMessage = migrator.downgrade(message, configuredAirbyteCatalog)
    writer.write(serializer.serialize(downgradedMessage))
    writer.newLine()
  }

  fun flush() {
    writer.flush()
  }

  fun close() {
    writer.close()
  }
}
